### ISSUE TITLE
(🐞) false negative with `dataclass` "non-default argument follows default argument"

### ISSUE THREAD

**Issue by KotlinIsland** [python/mypy#12137]
```py
from dataclasses import dataclass

@dataclass
class A:
    a: int | None = None

@dataclass
class B(A):
    a: int
    b: str
```

Mypy shows no error but this fails at runtime with `TypeError: non-default argument 'b' follows default argument`

**Comment by erictraut** [comment]
I don't see anywhere that this behavior is documented, but the runtime logic is clearly inheriting the default value from the base class even thought the child class doesn't provide a default value. This check was also missing in pyright, so I just added it.

**Comment by sobolevn** [comment]
Funny thing, this code works correctly:

```python
from dataclasses import dataclass

@dataclass
class A:
    a: int = 0

@dataclass
class B(A):
    b: str
```

But, `mypy` is confused about `a` redefinition in OP's example.

**Comment by noooonee** [comment]
I happened saw a similar question, can this answer fix your issue?

https://stackoverflow.com/a/69822584/1227329

**Comment by JonasFiechter** [comment]
Hello @noooonee. Adding `(kw_only=True)` into the decorator function Works for me, thanks
